File: injector/src/test/java/net/mready/inject/ModuleOverrideTest.java
package net.mready.inject;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class ModuleOverrideTest {
    @Test
    public void dependencyOverridenByModule() {
        Injector injector = Injector.with(new PlainStubOverrideModule());
        assertEquals(PlainStub.class, injector.instance(Plain.class).getClass());
    }


    @Test
    public void moduleOverwrittenBySubClass() {
        assertEquals("foo", Injector.with(new FooModule()).instance(String.class));
        assertEquals("bar", Injector.with(new FooOverrideModule()).instance(String.class));
    }

    public static class Plain {
    }

    public static class PlainStub extends Plain {

    }

    public static class PlainStubOverrideModule {
        @Provides
        public Plain plain(PlainStub plainStub) {
            return plainStub;
        }

    }

    public static class FooModule {
        @Provides
        String foo() {
            return "foo";
        }
    }

    public static class FooOverrideModule extends FooModule {
        @Provides
        @Override
        String foo() {
            return "bar";
        }
    }


}

File: photon/src/test/kotlin/net/mready/photon/ModuleOverrideTest.kt
package net.mready.photon

import org.junit.Assert.assertEquals
import org.junit.Test

class ModuleOverrideTest {
    @Test
    fun dependencyOverridenByModule() {
        val injector = Injector.with(PlainStubOverrideModule())
        assertEquals(PlainStub::class.java, injector.instance(Plain::class.java).javaClass)
    }

    @Test
    fun moduleOverwrittenBySubClass() {
        assertEquals("foo", Injector.with(FooModule()).instance(String::class.java))
        assertEquals("bar", Injector.with(FooOverrideModule()).instance(String::class.java))
    }

    open class Plain

    class PlainStub : Plain()

    class PlainStubOverrideModule {
        @Provides
        fun plain(plainStub: PlainStub): Plain {
            return plainStub
        }

    }

    open class FooModule {
        @Provides
        open fun foo(): String {
            return "foo"
        }
    }

    class FooOverrideModule : FooModule() {
        @Provides
        override fun foo(): String {
            return "bar"
        }
    }


}
